circuits/aes-gcm/nistgmul.circom

```diff
@@ -0,0 +1,268 @@
+pragma circom 2.1.9;
+
+include "utils.circom"; // xor
+include "circomlib/circuits/comparators.circom"; // isZero
+include "helper_functions.circom"; // bitwise right shift
+include "circomlib/circuits/mux1.circom"; // multiplexer
+include "../aes-ctr/utils.circom"; // xorbyte
+
+// Algorithm 1: X •Y
+// Input:
+// blocks X, Y.
+// Output:
+// block X •Y.
+// multiplication of two blocks in the binary extension field defined by the irreducible polynomial
+// 1 + X + X^2 + X^7 + X^128
+// computes a “product” block, denoted X •Y
+template NistGMulByte() {
+
+    signal input X[16];
+    signal input Y[16];
+    signal output out[16];
+
+    // Let R be the bit string 11100001 || 0120. Given two blocks X and Y
+    // byte 0xE1 is 11100001 in binary
+    // var R[16] = [0xE1, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00];
+
+    // 1. Let x0, x1...x127 denote the sequence of bits in X.
+    // 2. Let Z0 = 0128 and V0 = Y.
+    signal Z[129][16];
+    Z[0] <== [0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00];
+    /// State accumulator. ie. V[i] is V0 holding 16 bytes
+    signal V[129][16];
+    V[0] <== Y;
+    //       ⎧ Zi               if xi = 0;
+    //  Zi+1 ⎨ 
+    //       ⎩ Zi ⊕Vi           if xi =1.
+    //
+    // The V update is actually just gmulx (multiply binary polynomial by x)
+    //
+    //       ⎧ Vi >>1           if LSB1(Vi) = 0;
+    //  Vi+1 ⎨ 
+    //       ⎩ (Vi >>1) ⊕ R     if LSB1(Vi) =1.
+    //  
+    component bit[16];
+    component z_i_update[128];
+    component mulx[128];
+    component bytesToBits = BytesToBits(16);
+    bytesToBits.in <== X;
+    signal bitsX[16*8];
+    bitsX <== bytesToBits.out;
+    for (var i = 0; i < 128; i++) {
+        // log("i*8 + j", i*8 + j);
+        // z_i_update
+        z_i_update[i] = Z_I_UPDATE(16);
+        z_i_update[i].Z <== Z[i];
+        z_i_update[i].V <== V[i];
+        z_i_update[i].bit_val <== bitsX[i];
+        Z[i + 1] <== z_i_update[i].Z_new;
+
+        // mulx to update V
+        mulx[i] = Mulx(16);
```

Comment:
whoa, I can see how this introduces a lot of overhead.   I wonder if it's actually needed to be done per bit like this?